readthedocs/api/v3/mixins.py

```diff
@@ -0,0 +1,110 @@
+from django.contrib.auth.models import User
+from django.shortcuts import get_object_or_404
+from rest_framework.exceptions import NotFound
+
+from readthedocs.builds.models import Version
+from readthedocs.projects.models import Project
+
+
+class NestedParentObjectMixin:
+
+    # Lookup names defined on ``readthedocs/api/v3/urls.py`` when defining the
+    # mapping between URLs and views through the router.
+    PROJECT_LOOKUP_NAMES = [
+        'project__slug',
+        'projects__slug',
+        'superprojects__parent__slug',
+        'main_language_project__slug',
+    ]
+
+    VERSION_LOOKUP_NAMES = [
+        'version__slug',
+    ]
+
+    def _get_parent_object_lookup(self, lookup_names):
+        query_dict = self.get_parents_query_dict()
+        for lookup in lookup_names:
+            value = query_dict.get(lookup)
+            if value:
+                return value
+
+    def _get_parent_project(self):
+        slug = self._get_parent_object_lookup(self.PROJECT_LOOKUP_NAMES)
+        return get_object_or_404(Project, slug=slug)
+
+    def _get_parent_version(self):
+        project_slug = self._get_parent_object_lookup(self.PROJECT_LOOKUP_NAMES)
+        slug = self._get_parent_object_lookup(self.VERSION_LOOKUP_NAMES)
+        return get_object_or_404(
+            Version,
+            slug=slug,
+            project__slug=project_slug,
+        )
+
+
+class APIAuthMixin(NestedParentObjectMixin):
+
+    """
+    Mixin to define queryset permissions for ViewSet only in one place.
+
+    All APIv3 ViewSet should inherit this mixin, unless specific permissions
+    required. In that case, an specific mixin for that case should be defined.
+    """
+
+    def detail_objects(self, queryset, user):
+        # Filter results by user
+        # NOTE: we don't override the manager in User model, so we don't have
+        # ``.api`` method there
+        if self.model is not User:
+            queryset = queryset.api(user=user)
```

Comment:
We talked a little about this at the offsite. All of our code is using this and changing/removing it will require a bigger refactor.

I wrote a document explaining a little what each of these method do when I started touching auth in this API branch to match what we discussed: https://github.com/rtfd/readthedocs.org/pull/5624

Although, I think this is outside the scope of this PR. We will need to talk more and agree on a pattern to follow, though.